### PR TITLE
chore: refactor screens definitions and add more navigation events for screens

### PR DESCRIPTION
**Pull request by seanstrom**
### Summary

* This PR attempts to reorganise the definitions for all the app screens into separate groups
  * Ideally, these groups will help us selectively include more screens into the navigation tracking events.
  * Hopefully, grouping these screens will eventually lead to renaming screens to better describe their intent, and it could be nice to also use these groups for describing the "start" and "ends" of certain UX flows (from the code perspective).
  * Currently, this PR only extracts the existing definitions into separate groups, future work will begin leveraging these groups to help expand the navigation tracking.
* The existing screens have been reviewed and documented, and we have some **[notes for mobile app screens](https://www.notion.so/Mobile-App-Screens-Review-10c8f96fb65c80068f37f0591aa74e8b?pvs=4)** that can be used as a reference for future work.

---

Updated Summary

* This PR now decorates the screen definitions with extra metadata for configuring if the screen is tracked or not by analytics. The screens can also configure the name/id of the event that recorded when a user navigates to that screen.

### Testing notes

* There shouldn't be any breaking changes in this PR, but these changes do touch a core part of how the navigation system defines all of the screens that are accessible in the app. Meaning, that it could be important to do a regression test by navigating to "all the screens", or at least all of the major areas that could be affected like:
* Settings
* Wallet Settings
* Wallet screens for an account, the send flow, the bridge flow, etc
* Activity Center
* Scanning QR codes
* Contact Profiles
* Syncing screens
* Chat screens
* Communities screens

#### Platforms

- Android
- iOS

##### Functional

- Navigation to any screen in the app

### Steps to test

- Open Status
- Login into a profile
- Navigate through any part of the app

### Before and after screenshots comparison

WIP

status: ready 

### Risk

Described potential risks and worst case scenarios.

Tick **one**:
- [ ] Low risk: 2 devs MUST perform testing as specified above and attach their results as comments to this PR **before** merging.
- [x] High risk: QA team MUST perform additional testing in the specified affected areas **before** merging.

### CODE REVIEW

**Comment by ulisesmac** [comment]
Hi @seanstrom 

My thoughts on this PR:

I don't clearly see the problem we want to solve with this PR.

To me, some categories are not clearly defined: e.g. 
1. "Wallet" is a category with lots of screens, could they've been defined into smaller subcategories? what is the criteria to define a categroy? :thinking: 
3. IMO "Community" should also contain screens defined in the "Chat" category, since in the app we can get to a chat through a community too.

Since I don't see a clear definition of the categories, I think the current PR provides the same clarity as what we had achieved only with comments before  (i.e. `;; Wallet`, `;; Wallet Settings`, `;; Onboarding`, ...).

On the other hand, I think it'd be great that our screens were grouped, ideally (to me), our file structure and our screen groups should reflect in some way the navigation structure, so that adding/organizing/finding a screen is easy.

> future work will begin leveraging these groups to help expand the navigation tracking.

Is it possible for you to share the plans on how to achieve it?
 Why not to implement them in this PR or at least make a small proposal and, if liked/approved by others, continue the work?
 
 ------------
 
Sorry if I did a lot of questions, It's just a bit unclear to me the benefits from this PR.

Thanks @seanstrom !

**Comment by seanstrom** [comment]
Hey @ulisesmac, these are good questions, thanks for the review 🙏 

This PR is attempting to group the different screens into separate functions, but as you have mentioned they're not a precise enough abstraction over all of the screens. That being said the main benefit for grouping them into these "zones" will help with detecting which "kind" of screens is being displayed when we're tracking the navigation events.

For example, if we want to track more screen navigation events, but we want to be careful about which screens we're tracking, then we can reference a group of screens directly and extract the names of the screens. Here's some example code of what I'm thinking would be done in practice:

```clojure

;; screen definitions

(defn contact-screens
  []
  [{:name      :new-contact
    :options   {:sheet? true}
    :component add-new-contact/new-contact}

   {:name      :scan-profile-qr-code
    :options   options/dark-screen
    :component scan-profile-qr-page/view}

   {:name      :contact-profile
    :options   {:modalPresentationStyle :overCurrentContext}
    :component contact-profile/view}

   {:name      :share-contact
    :options   options/transparent-screen-options
    :component share-contact/view}])

(defn onboarding-screens
  []
  [{:name      :screen/onboarding.intro
    :options   {:theme :dark}
    :on-focus  [:onboarding/overlay-dismiss]
    :component intro/view}

   {:name      :screen/onboarding.syncing-results
    :options   {:theme :dark}
    :component syncing-results/view}])
      
;; tracking

(def ^:const view-ids-to-track
  (-> (concat (onboarding-screens) (contact-screens))
      (map :name)
      (into #{;; Tabs
              :communities-stack
              :chats-stack
              :wallet-stack})))

(defn track-view-id-event
  [view-id]
  (when (contains? view-ids-to-track view-id)
    (navigation-event (name view-id))))

(defn tracked-event
  [[event-name second-parameter]]
  (case event-name
    :profile/get-profiles-overview-success
    (user-journey-event app-started-event)

    :centralized-metrics/toggle-centralized-metrics
    (key-value-event "events.metrics-enabled" :enabled second-parameter)

    :set-view-id
    (track-view-id-event second-parameter)

    nil))

```

In this example we can see that the tracking code is now referencing the screen names without duplicating the names from the screens file, and it also helps us focus a specific subset of screens. And if we can focus on a subset then we can add additional metadata or logic for processing those navigation events, or even limit the amount of metadata when we think the event could represent sensitive information (like a wallet transaction or something).

> 
> To me, some categories are not clearly defined: e.g.
> 
> 1. "Wallet" is a category with lots of screens, could they've been defined into smaller subcategories? what is the criteria to define a categroy? 🤔
> 2. IMO "Community" should also contain screens defined in the "Chat" category, since in the app we can get to a chat through a community too.

1. Yeah absolutely! I also think we should define smaller subsets for the wallet screens, especially since we may want to track certain features like swaps or wallet-connect in different ways.
2. Yeah good point about the names of screens being shared across features. This is a tricky problem to contend with since we may want to identify when someone opens a community channel chat vs a messenger chat. There might be a way of doing this by inferring metadata from the app-state, but maybe we can create some extra screen aliases.
    * For example, maybe we can define a new screen like `:screen/community.channel`, and that would point to the same chat component, but we can reference the more distinct name when tracking the navigation event. 

> On the other hand, I think it'd be great that our screens were grouped, ideally (to me), our file structure and our screen groups should reflect in some way the navigation structure, so that adding/organizing/finding a screen is easy.
> 

Yeah this would be an ideal for me too, though I haven't attempted to re-structure things atm. For now I'm relying on the screens file to be source-of-truth for all the screen names.

> Is it possible for you to share the plans on how to achieve it?
> Why not to implement them in this PR or at least make a small proposal and, if liked/approved by others, continue the work?
> 

Yeah we can continue to improve this PR to have better groups or aliases for the screens, atm I decided to try something minimal with low-risk of breaking things, but we can continue to refine this approach in this PR or as we go. Though I did plan to experiment with the tracking code in a separate PR when we start expanding the navigation events, but I suppose we can at least replace the onboarding detection and possibly add more screens to the tracking logic too.

---

I hope this commentary helps clarify things, and feel free to share more too. I imagine that this will be a collective effort for managing screens over time, so we should have something that we can all maintain and enjoy 🙌

**Comment by ilmotta** [comment]
@seanstrom: Grouping screens in different vars is a nice improvement, I see no harm in that, but using that as a proxy for meaning _trackable_ I think is mixing orthogonal concerns. It will cause trouble later because we will need to regroup screens to indirectly track/not track them.

A more direct solution, which also doesn't rely on hierarchies is: we can add a flag to each navigation screen map in case it is allowed to be tracked. Then, in the tracking namespace, similar to what you shared above, we statically build a set of `view-ids-to-track`. We could do the opposite to build a blocklist too, but the allowlist would be slightly safer to avoid the case of tracking new screens by mistake (a dev can easily forget to disable if the default is to track).

The code would be trivial:

```clojure
{:name      :screen/xyz
 :track?    true ; => ********* ADDED
 :options   {:insets {:top? true}}
 :component ...}

;; In the ns status-im.contexts.centralized-metrics.tracking
(def ^:const view-ids-to-track
  (into #{;; Tabs
          :communities-stack
          :chats-stack
          :wallet-stack}
        (->> (screens/screens)
             (filter :track?)
             (map :name))))
```

-----

About the discussion of code organization of the screens, I kind of prefer a central place than spreading them over directories, which would make finding them even more difficult, just 2 cents.

One thing that's been on my radar for a long time was to extract each screen map into a separate var because that would instantly give us reference finding, unused screens, and jumping to source. Would be super helpful to everybody if we did this 🙊

**Review by ulisesmac — APPROVED** [review]
@ilmotta @seanstrom 
After considering your points, I'm approving this PR now since I agree on the benefit of tracking data.

And, just to clarify, ideally, I'd like to have a Clojure data structure and a file hierarchy that refelects the navigation:

e.g. for files
```
root
|
|--- Onboarding
         |
         |--- Welcome
         |--- Sign-in
|
|--- Wallet
        |
        |-- Home
|
|--- Communities
|--- Messsages
|--- Browser
```

Something similar to [Expo Router (used for React Native)](Introduction to Expo Router)

> It brings the best file-system routing concepts from the web to a universal application — allowing your routing to work across every platform. When a file is added to the app directory, the file automatically becomes a route in your navigation.


So it'd impact our  `(defn screens [] ...)`, we could even create a macro to do the job, but we first need to make clear what is the best data structure.

**Comment by seanstrom** [comment]
@ilmotta @ulisesmac thanks for the feedback 🙏 

I agree that we should add more data to each screen for configuring `:trackable?`. That idea makes me think we could even create a `:metrics` key for each screen that could contain as much metadata as we need over time.

File: src/status_im/contexts/centralized_metrics/events.cljs
Comment (seanstrom) [inline src/status_im/contexts/centralized_metrics/events.cljs]: Here we allow for processing a vector of events provided by the tracking function. This is useful because we can additional event data to the existing event data that's already being recorded.

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (seanstrom) [inline src/status_im/contexts/centralized_metrics/tracking.cljs]: Here we're creating a new event for naming the individual screens as their own event. This kind of abstraction kind be done inside Mixpanel with Custom Events, but we can also decide whether we want to emit a bespoke event for certain screens too.

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (seanstrom) [inline src/status_im/contexts/centralized_metrics/tracking.cljs]: Here we're referencing all the registered screens and determining if the screen is configured to be tracked.

Comment (seanstrom) [reply src/status_im/contexts/centralized_metrics/tracking.cljs]: We also emit `screen-events` with the navigation event, but we should decide whether we want these screen events, or maybe if we only want them for a subset of screen.

**Comment by seanstrom** [comment]
Okay I've made some changes based on some experiments and your feedback. Now we can configure if a screen is tracked by tagging the screen definition with some extra metadata, and we can also alias the event-id / event name that gets sent to Mixpanel too. We also have the ability to send screen specific events (if we want to), please let me know what you think when you have a chance 🙏 

cc @ilmotta @ulisesmac

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (ilmotta) [inline src/status_im/contexts/centralized_metrics/tracking.cljs]: The default value `false` seems unusual/unnecessary because the falsey result from `get-in` should be sufficient.

File: status-mobile.code-workspace
Comment (ilmotta) [inline status-mobile.code-workspace]: Just checking, is this file being intentionally added? Good for other devs too?

File: src/status_im/contexts/centralized_metrics/events.cljs
Comment (ilmotta) [inline src/status_im/contexts/centralized_metrics/events.cljs]: @seanstrom, `coll?` will return true if `event` is a map, for example. Maybe here we should only use `vector?` to be extra precise?

Could be an `if` instead of `cond`. `doseq` is preferred to process with side-effects and ignore results.

```clojure
(if (vector? event)
  (doseq [e event]
    (native-module/add-centralized-metric e))
  (native-module/add-centralized-metric event))
```

**Comment by ilmotta** [comment]
> Something similar to [Expo Router (used for React Native)](Introduction to Expo Router)
> 
> > It brings the best file-system routing concepts from the web to a universal application — allowing your routing to work across every platform. When a file is added to the app directory, the file automatically becomes a route in your navigation.
> 
> So it'd impact our `(defn screens [] ...)`, we could even create a macro to do the job, but we first need to make clear what is the best data structure.

@ulisesmac This reminds of the way Rails made _convention over configuration_ famous more than 15 years ago 🚋 I tend to prefer more explicit configuration, rather than implicit because it's simpler to accommodate for pretty much any use case. But I don't have any strong preference, but other CCs may have.

File: src/status_im/navigation/screens.cljs
Comment (ilmotta) [inline src/status_im/navigation/screens.cljs]: I'm on the fence about adding `:flow` now because we aren't using this data yet, or maybe I missed in the diff. Given that `event.id` keyword already has a useful qualification, in the analytics tool this may be sufficient for filtering purposes (not sure).

**Review by ilmotta — APPROVED** [review]
I'm leaving a few minor suggestions @seanstrom. As usual, great work with the thoughtful implementation.

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (ilmotta) [inline src/status_im/contexts/centralized_metrics/tracking.cljs]: The binding `screens-by-name` seems unnecessary.

File: status-mobile.code-workspace
Comment (seanstrom) [reply status-mobile.code-workspace]: ooops, this is my special file 🙈

**Review by seanstrom — COMMENTED** [review]
Self-Review 📓

File: src/status_im/contexts/centralized_metrics/events.cljs
Comment (seanstrom) [reply src/status_im/contexts/centralized_metrics/events.cljs]: Ah good call for using `doseq` 👍 

I think you're right that we can remove the `coll?` check. Additionally, I think we can also support seqs by using `seq?` since we might return a sequence or vector from the tracking function, how does that sound?

File: src/status_im/navigation/screens.cljs
Comment (seanstrom) [reply src/status_im/navigation/screens.cljs]: Happy to remove these values 👍 

For context: currently `:flow` is not used for anything, so we do not need that property, it's mostly there as a way to categorise the screens without relying too much on the variable names in the file.

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (clauxx) [inline src/status_im/contexts/centralized_metrics/tracking.cljs]: `get-in` will return a falsy value already, so no need for the default `false`

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (clauxx) [inline src/status_im/contexts/centralized_metrics/tracking.cljs]: what is the difference between `view-id` and `screen-id`. Is there any value in tracking both?

File: src/status_im/navigation/screens.cljs
Comment (clauxx) [inline src/status_im/navigation/screens.cljs]: Is the event id for finding the event in the dashboard? Can we just use the name of the screen, just to avoid creating many "slightly different" identifiers?

File: src/status_im/navigation/screens.cljs
Comment (clauxx) [reply src/status_im/navigation/screens.cljs]: +1 for removing them, since we are already categorizing them with vars

**Review by clauxx — APPROVED** [review]

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (seanstrom) [reply src/status_im/contexts/centralized_metrics/tracking.cljs]: `:screen-id` (recently renamed to `:screen-name`) is the string formatted like this `screen/profile.settings`. And `viewId` would be the string `profile.settings`. `screen-id` is mostly here to give an exact match on the keyword used for identifying the screen.

File: src/status_im/navigation/screens.cljs
Comment (seanstrom) [reply src/status_im/navigation/screens.cljs]: So atm the screens partially follow a naming pattern like: `domain.section` or maybe `domain.sub-domain.section` (if necessary). But we don't really do this for all the screens atm, so this event-id sorta decouples the current names from the reported names, which allows us to send the more ideal screen name for now. 

Ideally, the names of the screens would be renamed to match the event-id, but doing that for all the screens seemed a bit risky atm. So the introduction of the event-id is sorta a migration step.

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (clauxx) [reply src/status_im/contexts/centralized_metrics/tracking.cljs]: do we need both? Just wondering if we're not sending redundant data.

Comment (seanstrom) [reply src/status_im/contexts/centralized_metrics/tracking.cljs]: Having both is likely redundant, and I would be in favour of sending one of them, I think the value for `screen-id` is more accurate, but I've left both for now for backwards compatibility reasons.

If we get green-light to use the screen-id value, then we can remove the other one, but for the moment we're extending the data to avoid a breaking change.

Comment (clauxx) [reply src/status_im/contexts/centralized_metrics/tracking.cljs]: thanks for clearing it up :+1:

**Review by clauxx — APPROVED** [review]
Looks good!

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (shivekkhurana) [inline src/status_im/contexts/centralized_metrics/tracking.cljs:45]: Thank you. This approach is much better than what I did with manually adding the screen names.

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (shivekkhurana) [inline src/status_im/contexts/centralized_metrics/tracking.cljs:47]: Why do we need this ? Why not use `:track?` on the other views too ? 

Do `stacks` behave differently from a `screen`?

File: src/status_im/navigation/screens.cljs
Comment (shivekkhurana) [inline src/status_im/navigation/screens.cljs]: Why do we need to `defn` and then call it later ? Why not just `def` and screens can just be:

```clj
(def chat-screens
  ...)

(concat 
  old-screens/screens 
  chat-screens ...)
```

File: src/status_im/contexts/centralized_metrics/events.cljs
Comment (shivekkhurana) [inline src/status_im/contexts/centralized_metrics/events.cljs:20]: In what cases will event be a `seq` ?

**Review by shivekkhurana — COMMENTED** [review]
The PR looks good but I have asked for some clarifications.

File: src/status_im/navigation/screens.cljs
Comment (shivekkhurana) [reply src/status_im/navigation/screens.cljs]: I also feel we should not add it here. Flow can be `inferred` from the data.

File: src/status_im/contexts/centralized_metrics/tracking.cljs
Comment (seanstrom) [reply src/status_im/contexts/centralized_metrics/tracking.cljs:47]: yup yup, stacks seem to be set manually by some other UI logic, the screen that contains them is called `:shell-stack` I think.

File: src/status_im/contexts/centralized_metrics/events.cljs
Comment (seanstrom) [reply src/status_im/contexts/centralized_metrics/events.cljs:20]: Originally I added support for returning a vector of events, but then I thought that someday some will a `map` function on a vector, and that would turn the vector into a `seq`.

File: src/status_im/navigation/screens.cljs
Comment (seanstrom) [reply src/status_im/navigation/screens.cljs]: Yeah I think it's possible that they can all be turned into defs, at the time I assumed there was a reason for `screens` being a function, but maybe that's for legacy reasons.

cc @ilmotta

Comment (ilmotta) [reply src/status_im/navigation/screens.cljs]: As far as I remember, one of the reasons to choose functions is that we sometimes may prefer to defer execution to be triggered at a later stage, so that we impact less the already slow(ish) startup of the app (in some devices). But this idea doesn't seem to make much sense for this code, so probably unrelated.

Might be just personal preference. I've seen devs who prefer to use functions, give names to them, and then treat the vars as global state, which means they might prefer to put as little code as possible in the var definition directly. A matter of taste maybe.

**Comment by mariia-skrypnyk** [comment]
Hey @seanstrom !

Is PR ready to be tested? Or you plan to do more changes?

**Comment by seanstrom** [comment]
Hey @mariia-skrypnyk i think the PR is ready, but there are some other PRs that will cause me to do a rebase, so we may want to wait for those to merge first.

Specifically PR https://github.com/status-im/status-mobile/pull/21379 is scheduled to merge before this PR

**Comment by mariia-skrypnyk** [comment]
@seanstrom thanks, so should I start testing after merg of https://github.com/status-im/status-mobile/pull/21379 ?

**Comment by seanstrom** [comment]
@mariia-skrypnyk yup yup, after it merges and I rebase to include the changes in this PR

**Comment by mariia-skrypnyk** [comment]
Please, tag me when it will be ready for me 🙏

File: shadow-cljs.edn
Comment (seanstrom) [inline shadow-cljs.edn:99]: Here we're using a shadow-cljs feature for conditionally requiring the metrics interceptor logic for all the builds except the component-tests. This is because requiring the metrics interceptor in the component tests will cause several JS imports, which is caused by the metrics interceptor requiring the screens namespace which eventually imports all the screen imports for third-party modules.

So instead we configured shadow-cljs to and the interceptor namespace to exclude loading the metrics interceptor to avoid the extra imports that would require extra JS mocking for the component specs.

**Comment by seanstrom** [comment]
@mariia-skrypnyk I think this PR is now ready for testing 🙏

**Comment by mariia-skrypnyk** [comment]
Hey @seanstrom !

Thanks for pinging me up.
I've started to test your PR and run tests.
A lot of them fall because of not fetching communities but one of them was because no Status on Discover community screen. It didn't work for some time but it did after a while.

**Comment by seanstrom** [comment]
@mariia-skrypnyk thanks for testing the PR 🙏 

Though it seems we'll need to wait for https://github.com/status-im/status-mobile/pull/21280 to merge before we can continue with PR

**Comment by pavloburykh** [comment]
> @mariia-skrypnyk thanks for testing the PR 🙏
> 
> Though it seems we'll need to wait for #21280 to merge before we can continue with PR

Hey @seanstrom I have moved https://github.com/status-im/status-mobile/pull/21280 forward. Once @vkjr merges it you can rebase your branch.

**Comment by vkjr** [comment]
@pavloburykh, it is merged, thanks!

**Comment by mariia-skrypnyk** [comment]
Hey @seanstrom !

https://github.com/status-im/status-mobile/pull/21280 is merged.

What is our next step?

**Comment by seanstrom** [comment]
Hey @mariia-skrypnyk 👋 

I've pushed up the latest changes into this PR, so we should be ready to continue testing.
I think the main areas that could be affected by my changes would be navigating to the screens in the app.
I imagine we need to manually check that each screen is still able to be reached by navigation, and if they are not then that's likely a bug caused by this PR.

So far I've manually tested a lot of the screens on my simulator, but there are few areas that could use some extra checking. For example:
* The screens related to communities or token gated communities
* The screens related to wallet-swap, wallet-bridge, wallet-connect
* And some screens for the emoji-picker and pdf-viewer (not sure how to use these features)

The majority of all the other screens I've been able to verify that we can still reach them through navigation. This would include screens for things like onboarding, syncing, chats, contacts, settings, QR scanners and codes. Notably, I have not tested the keycard screens atm. I've also verified some of the wallet screens for creating / editing an account and key-pair. And I've also looked at the different flows for importing a key-pair. 

What you think about this testing strategy? Let me know if there's anything else I can do to help 🙏

**Comment by mariia-skrypnyk** [comment]
Thanks @seanstrom !

Looks great! Will check communities and wallet features tomorrow.
Also will try to look at a keycard.

**Comment by mariia-skrypnyk** [comment]
Hey @seanstrom !

I've checked PR including your points:

- Wallet (send, swap, bridge, WC)
- Communities (had no chance to check roken gated as their behaviour is unstable and no chance to get into it even if you have needed tokens)
- Emoji in chats, reactions
- Upgrade from v1 to v2 using keycard

Everything looks good and PR can be merged.

**Pull request merged**
